core/src/main/scala/org/apache/spark/SparkStatusTracker.scala

```diff
@@ -120,4 +120,8 @@ class SparkStatusTracker private[spark] (sc: SparkContext, store: AppStatusStore
         exec.memoryMetrics.map(_.totalOnHeapStorageMemory).getOrElse(0L))
     }.toArray
   }
+
+  def getAppStatusStore: AppStatusStore = {
+    store
+  }
```

Comment:
We dont need to use AppStatusStore for this PR, but can directly fetch the stage/task state from TSM.

Reply:
It might  not be able to fetch the stage/task state (metrics) from TaskSetManager directly.   Because the tasks  are running in the executors, and the executors report (heartbeat) their task metrics to the driver,  the driver gets the Heartbeat and  posts a SparkListenerExecutorMetricsUpdate to the listenerBus, and AppStatusListener.onExecutorMetricsUpdate will get the  ExecutorMetrics and update them. So we may get the task Metrics from AppStatusStore.

Reply:
This is a good point - for running tasks, currently we are not updating TSM with the details - since scheduler did not care for this, until now.
We can introduce the specific subset of metrics we want to track, and update TaskInfo with it - private to scheduler package.

Thoughts ?
+CC @Ngone51

Reply:
This sounds like we'd duplicate the (part of) metrics in Spark and the metrics become scattered instead of being centralized in one place.

Reply:
There are a few aspects here:

a) Should we track metrics in scheduler ?

Task metrics are part of `TaskInfo` in scheduler - a view of this state is exposed using the listener, and is used by the UI (`AppStatusStore`, etc): but source of truth is the scheduler.
This is maintained within scheduler only for completed states currently.

In order to make the UI more responsive to task progress, metrics also sent via executor heartbeat (SPARK-2099, etc) - so that users can see the inprogress updates to metrics for running tasks.
This is not tracked at the scheduler, since we did not need any of this information (and directly fire the `SparkListenerExecutorMetricsUpdate` event).
Now that we do, we can leverage the heartbeat to update relevant state.

(Note, I am not proposing to maintain all `accumulables`, just the once we need - see next pls).

b) Should we materialize specific metrics ?

I would look at it as an optimization - to materialize the specific subset of metrics we need.
For completed tasks, this is already available - but will require a scan through `accumulables` (and so a performance gain by materializing them).
For in progress tasks, this is currently not maintained in scheduler - and so tracking them during heatbeat would allow scheduler to leverage the state.

c) Going to `AppStatusStore`, etc.

There are two things which impact this design choice - the cost and reliability.

With speculation enabled, we check for speculative tasks even 100ms. Since this is within the scheduler lock, we should ensure it completes quickly (as other things will be blocked). When the total number of task increases, the performance of `InMemoryStore` degrades (and also has a bad gc impact) [1] - which results in scheduler/driver getting impacted.

From a reliability point of view, the listener subsystem can and does drop events - so depending on it is introducing error into scheduler decision : which can be avoided by tracking this in scheduler itself.


Hopefully I am not missing anything, please let me know your thoughts @Ngone51 !


[1] For the current usecases for this store, this is fine (infrequently powering spark UI)

Reply:
This sounds good to me. Just adds one note:

> For completed tasks, this is already available - but will require a scan through accumulables (and so a performance gain by materializing them).
> For in progress tasks, this is currently not maintained in scheduler - and so tracking them during heatbeat would allow scheduler to leverage the state.

I think this essentially means we'll have intermediate `accumulables` for `TaskInfo` rather than only final `accumulables` for the completed tasks as what we have today. And we'll have to track all tasks since the completed tasks were inprogress tasks ever.

Reply:
> I think this essentially means we'll have intermediate accumulables for TaskInfo rather than only final accumulables for the completed tasks as what we have today

Materializing the subset of required values was an optimization to this - since `_accumulables` is a `Seq` and the scan would be done repeatedly (we only need a small subset of input/shuffle related metrics to determine progress, while the total set can be fairly large).

> And we'll have to track all tasks since the completed tasks were inprogress tasks ever.

For completed tasks, we are already tracking this.
For in-progress tasks, we are not - and will need to be added.
For tasks which are yet to start, this would be empty.

Reply:
OK, `intermediate accumulables` is not accurate. We'd have a subset of required metrics instead. 

So for a running task, IIUC, we'll still have to scan all the `accumulables` on every heartbeat in order to update the required metrics, right?

Reply:
Yes, you are right - a scan over each heartbeat (once every 10s per executor).
Btw, if we can make this cheaper, I am all for it !
What I want to make sure is that the speculation check is as fast as possible

Reply:
@mridulm @Ngone51 Thanks for your suggestions, I submitted a new commit with a small change in the AppStatusListener to improve tracking metrics for successful and in-progress tasks, which might be cheaper to evaluate the efficiency of task processing.

Why do we use an existing AppStatusListener instead of creating a new Listener?  Because using the AppStatusListener can reuse some data structures and execution logic for us.

Reply:
No hurries, take your time - pls ping me when you are done. I will try to take a look at the changes later this week. Thx for pushing through on the feature !